File: systori/dart/lib/src/editor/stickyheader.dart
import 'dart:html';
import 'dart:math' as math;
import 'package:quiver/iterables.dart';
import 'editor.dart';


class StickyHeader {

    Job job;

    static final bool DEBUG = false;
    static final int TOP = 52;
    List<Group> oldFixedStack = [];
    List<Group> oldSlidingStack = [];
    DivElement ts, bs;

    StickyHeader(this.job) {
        window.onScroll.listen(handleScroll);
        if (DEBUG) {
            ts = document.createElement('div');
            ts.style.backgroundColor = 'green';
            bs = document.createElement('div');
            bs.style.backgroundColor = 'red';
            for (var s in [ts, bs]) {
                s.style.position = 'fixed';
                s.style.width = '1px';
                s.style.height = '2px';
                s.style.zIndex = '1000';
                job.append(s);
            }
            updateSensorMarkers();
        }
    }

    updateSensorMarkers([int bottom]) {
        if (DEBUG) {
            ts.style.top = '${TOP}px';
            ts.style.left = '${job.offsetLeft - 2}px';
            bs.style.left = '${job.offsetLeft - 2}px';
            if (bottom != null)
                bs.style.top = '${bottom}px';
            else
                bs.style.top = '0px';
        }
    }

    int stackHeight(List<Group> stack) {
        var height = 0;
        for (var model in stack) {
            height += model.children[0].children[0].clientHeight;
        }
        return height;
    }

    handleScroll(Event e) {

        if (DEBUG) print(''.padRight(10, '-'));

        Element topElement = document.elementFromPoint(job.offsetLeft, TOP);
        Group topGroup = findGroup(topElement);
        if (topGroup == null) {
            /* short circuit */
            drawStacks([], []);
            updateSensorMarkers();
            if (DEBUG) {
                print("Clearing header.");
                print("".padRight(10, '+'));
            }
            return;
        }

        List<Group> topStack = makeGroupStack(topGroup);
        var topStackHeight = stackHeight(topStack);

        if (DEBUG) print('Top Sensor: ${topGroup.name.text}');

        var middleSensor = (topStackHeight/2).round() + TOP;
        Element middleElement = document.elementFromPoint(job.offsetLeft, middleSensor);
        Group middleGroup = topElement == middleElement ? topGroup : findGroup(middleElement);

        var bottomSensor = topStackHeight + TOP;
        Element bottomElement = document.elementFromPoint(job.offsetLeft, bottomSensor);
        Group bottomGroup = middleElement == bottomElement ? middleGroup : findGroup(bottomElement);

        if (DEBUG) print('Btm Sensor: ${bottomGroup?.name?.text}');

        if (topGroup == bottomGroup || (topGroup == middleGroup && bottomGroup == null)) {
            // If top and bottom (or middle) sensors both see the same thing,
            // just drawing the topStack as a fixed header without any further
            // processing or merging.
            if (DEBUG) {
                print('Top: ${topStack.map((g)=>g.name.text).toList()}');
                print('top == bottom');
            }
            drawStacks(topStack, []);
            updateSensorMarkers(bottomSensor);
        } else if (middleGroup != topGroup && middleGroup != bottomGroup) {
            // If the middle sensor found a distinct group then make that the new
            // top.
            if (DEBUG) print('using middle group as top');
            processStack(makeGroupStack(middleGroup));
        } else if (topGroup == bottomGroup.parent) {
            // If bottom group is a child of the top group then optimistically
            // make the bottomGroup the new top.
            if (DEBUG) print('using bottom group as top');
            processStack(makeGroupStack(bottomGroup));
        } else {
            // No special cases found, just process the top stack.
            if (DEBUG) print('using top group');
            processStack(topStack, bottomGroup);
        }

        if (DEBUG) print("".padRight(10, '+'));
    }

    processStack(List<Group> topStack, [Group bottomGroup]) {

        int bottomSensor = stackHeight(topStack) + TOP;
        if (bottomGroup == null) {
            bottomGroup = findGroup(document.elementFromPoint(job.offsetLeft, bottomSensor));
        }
        List<Group> bottomStack = makeGroupStack(bottomGroup);
        List<Group> fixed = [], sliding = [];

        mergeStacks(topStack, bottomStack, fixed, sliding);

        if (DEBUG) {
            print('Top: ${topStack.map((g)=>g.name.text).toList()}');
            print('Btm: ${bottomStack.map((g)=>g.name.text).toList()}');
            print('Fixed: ${fixed.map((g)=>g.name.text).toList()}');
            print('Slide: ${sliding.map((g)=>g.name.text).toList()}');
        }

        int slidingStart;
        if (sliding.isNotEmpty) {
            var fixedBottom = fixed.last.children[0].children[0]
                .getBoundingClientRect()
                .bottom;
            var rec = bottomGroup.children[0].getBoundingClientRect();
            slidingStart = rec.top.round() - 1;
            if (fixedBottom >= slidingStart) {
                if (DEBUG) print('fixed = bottomStack');
                fixed = bottomStack;
            }
        }

        drawStacks(fixed, sliding, slidingStart);
        updateSensorMarkers(bottomSensor);

    }

    mergeStacks(List<Group> topStack, List<Group> bottomStack, List<Group> fixed, List<Group> sliding) {
        var longest = math.max(topStack.length, bottomStack.length);
        for (int i = 0; i < longest; i++) {
            var topModel = i < topStack.length ? topStack[i] : null;
            var bottomModel = i < bottomStack.length ? bottomStack[i] : null;
            if (topModel == bottomModel) {
                fixed.add(topModel);
            } else {
                if (topModel != null) {
                    sliding.insert(0, topModel);
                } else if (bottomModel != null) {
                    fixed.add(bottomModel);
                }
            }
        }
    }

    drawStacks(List<Group> newFixedStack, List<Group> newSlidingStack, [int slidingStart]) {

        for (var group in concat([oldFixedStack, oldSlidingStack])) {
            if (!newFixedStack.contains(group) && !newSlidingStack.contains(group)) {
                var editor = group.children[0];
                var row = editor.children[0];
                if (row.classes.contains('sticky-header')) {
                    row.style.top = null;
                    row.style.width = null;
                    row.classes.remove('sticky-header');
                    editor.children.removeAt(1);
                    if (DEBUG) print('"${group.name.text}".removeAt(1)');
                    row.classes.remove('sliding');
                }
            }
        }

        for (var group in concat([newFixedStack, newSlidingStack])) {
            var editor = group.children[0];
            var row = editor.children[0];
            if (!row.classes.contains('sticky-header')) {
                if (DEBUG) print('"${group.name.text}".insertBefore(document.createElement(div), ...)');
                var spacer = document.createElement('div');
                spacer.style.height = '${row.clientHeight}px';
                row.style.width = '${row.clientWidth}px';
                row.classes.add('sticky-header');
                editor.insertBefore(spacer, editor.children[1]);
            }
        }

        var offset = TOP - 1;
        for (var model in newFixedStack) {
            var editor = model.children[0];
            var row = editor.children[0];
            row.style.top = '${offset}px';
            offset += row.clientHeight - 1;
            row.classes.toggle('sliding', false);
        }

        for (var model in newSlidingStack) {
            var editor = model.children[0];
            var row = editor.children[0];
            slidingStart -= row.clientHeight;
            row.style.top = '${slidingStart}px';
            row.classes.toggle('sliding', true);
        }

        oldFixedStack = newFixedStack;
        oldSlidingStack = newSlidingStack;
    }

    Group findGroup(Element element) {
        while (element != null) {
            if (element is Group) {
                return element;
            }
            element = element.parent;
        }
        return null;
    }

    List<Group> makeGroupStack(Group group) {
        List<Group> newStack = [];
        if (group != null) {
            newStack.insert(0, group);
            while (group != null && group is! Job) {
                group = group.parent;
                if (group is Group) {
                    newStack.insert(0, group);
                }
            }
        }
        return newStack;
    }
}

File: systori/static/js/editor/stickyheader.js
/**
 * Represents a sticky header.
 */
class StickyHeader {
    /**
     * Creates a StickyHeader instance.
     *
     * @param {Job} job - The Job instance associated with this header.
     */
    constructor(job) {
        /** @type {Job} */
        this.job = job;

        /** @type {boolean} */
        this.DEBUG = false;

        /** @type {number} */
        this.TOP = 52;

        /** @type {Array<Group>} */
        this.oldFixedStack = [];

        /** @type {Array<Group>} */
        this.oldSlidingStack = [];

        /** @type {HTMLElement} */
        this.ts = null;

        /** @type {HTMLElement} */
        this.bs = null;

        window.addEventListener('scroll', this.handleScroll.bind(this));
        // Bind the methods to the class instance
        this.updateSensorMarkers = this.updateSensorMarkers.bind(this);
        this.handleScroll = this.handleScroll.bind(this);

        if (this.DEBUG) {
            this.ts = document.createElement('div');
            this.ts.style.backgroundColor = 'green';
            this.bs = document.createElement('div');
            this.bs.style.backgroundColor = 'red';

            [this.ts, this.bs].forEach(s => {
                s.style.position = 'fixed';
                s.style.width = '1px';
                s.style.height = '2px';
                s.style.zIndex = '1000';
                this.job.append(s);
            });

            this.updateSensorMarkers();
        }
    }
    /**
     * Updates the positions of sensor markers.
     * @param {number} [bottom] - The bottom position for the sensor marker. Optional.
     */
    updateSensorMarkers(bottom) {
        if (this.DEBUG) {
            this.ts.style.top = `${this.TOP}px`;
            this.ts.style.left = `${this.job.offsetLeft - 2}px`;
            this.bs.style.left = `${this.job.offsetLeft - 2}px`;

            if (bottom !== undefined) {
                this.bs.style.top = `${bottom}px`;
            } else {
                this.bs.style.top = '0px';
            }
        }
    }
    /**
     * Calculates the total height of a stack of Group elements.
     * @param {Array<Group>} stack - An array of Group elements.
     * @returns {number} The total height of the stack.
     */
    stackHeight(stack) {
        let height = 0;
        for (let model of stack) {
            height += model.children[0].children[0].clientHeight;
        }
        return height;
    }
    /**
     * Handles the scroll event.
     * @param {Event} e - The event object.
     */
    handleScroll(e) {
        if (this.DEBUG) console.log('-'.repeat(10));

        let topElement = document.elementFromPoint(this.job.offsetLeft, this.TOP);
        let topGroup = this.findGroup(topElement);
        if (!topGroup) {
            this.drawStacks([], []);
            this.updateSensorMarkers();
            if (this.DEBUG) {
                console.log("Clearing header.");
                console.log('+'.repeat(10));
            }
            return;
        }

        let topStack = this.makeGroupStack(topGroup);
        let topStackHeight = this.stackHeight(topStack);

        if (this.DEBUG) console.log(`Top Sensor: ${topGroup.name.text}`);

        let middleSensor = Math.round(topStackHeight / 2) + this.TOP;
        let middleElement = document.elementFromPoint(this.job.offsetLeft, middleSensor);
        let middleGroup = topElement === middleElement ? topGroup : this.findGroup(middleElement);

        let bottomSensor = topStackHeight + this.TOP;
        let bottomElement = document.elementFromPoint(this.job.offsetLeft, bottomSensor);
        let bottomGroup = middleElement === bottomElement ? middleGroup : this.findGroup(bottomElement);

        if (this.DEBUG) console.log(`Btm Sensor: ${bottomGroup?.name?.text}`);

        if (topGroup === bottomGroup || (topGroup === middleGroup && !bottomGroup)) {
            if (this.DEBUG) {
                console.log(`Top: ${topStack.map(g => g.name.text).join(', ')}`);
                console.log('top == bottom');
            }
            this.drawStacks(topStack, []);
            this.updateSensorMarkers(bottomSensor);
        } else if (middleGroup !== topGroup && middleGroup !== bottomGroup) {
            if (this.DEBUG) console.log('using middle group as top');
            this.processStack(this.makeGroupStack(middleGroup));
        } else if (topGroup === bottomGroup.parent) {
            if (this.DEBUG) console.log('using bottom group as top');
            this.processStack(this.makeGroupStack(bottomGroup));
        } else {
            if (this.DEBUG) console.log('using top group');
            this.processStack(topStack, bottomGroup);
        }

        if (this.DEBUG) console.log('+'.repeat(10));
    }
    /**
     * Processes the top and bottom stacks, and merges them into fixed and sliding stacks.
     * @param {Array<Group>} topStack - The top stack of Group elements.
     * @param {Group} [bottomGroup] - The bottom Group element, optional.
     */
    processStack(topStack, bottomGroup) {
        let bottomSensor = this.stackHeight(topStack) + this.TOP;
        if (!bottomGroup) {
            bottomGroup = this.findGroup(document.elementFromPoint(this.job.offsetLeft, bottomSensor));
        }
        let bottomStack = this.makeGroupStack(bottomGroup);
        let fixed = [], sliding = [];

        this.mergeStacks(topStack, bottomStack, fixed, sliding);

        if (this.DEBUG) {
            console.log(`Top: ${topStack.map(g => g.name.text).join(', ')}`);
            console.log(`Btm: ${bottomStack.map(g => g.name.text).join(', ')}`);
            console.log(`Fixed: ${fixed.map(g => g.name.text).join(', ')}`);
            console.log(`Slide: ${sliding.map(g => g.name.text).join(', ')}`);
        }

        let slidingStart;
        if (sliding.length > 0) {
            let fixedBottom = fixed[fixed.length - 1].children[0].children[0]
                .getBoundingClientRect()
                .bottom;
            let rec = bottomGroup.children[0].getBoundingClientRect();
            slidingStart = Math.round(rec.top) - 1;
            if (fixedBottom >= slidingStart) {
                if (this.DEBUG) console.log('fixed = bottomStack');
                fixed = bottomStack;
            }
        }

        this.drawStacks(fixed, sliding, slidingStart);
        this.updateSensorMarkers(bottomSensor);
    }
    /**
     * Merges the top and bottom stacks into fixed and sliding stacks.
     * @param {Array<Group>} topStack - The top stack of Group elements.
     * @param {Array<Group>} bottomStack - The bottom stack of Group elements.
     * @param {Array<Group>} fixed - The fixed stack to populate.
     * @param {Array<Group>} sliding - The sliding stack to populate.
     */
    mergeStacks(topStack, bottomStack, fixed, sliding) {
        let longest = Math.max(topStack.length, bottomStack.length);
        for (let i = 0; i < longest; i++) {
            let topModel = i < topStack.length ? topStack[i] : null;
            let bottomModel = i < bottomStack.length ? bottomStack[i] : null;
            if (topModel === bottomModel) {
                fixed.push(topModel);
            } else {
                if (topModel !== null) {
                    sliding.unshift(topModel); // insert at beginning
                } else if (bottomModel !== null) {
                    fixed.push(bottomModel);
                }
            }
        }
    }
    /**
     * Draws the new fixed and sliding stacks, updating the position and appearance of elements.
     * @param {Array<Group>} newFixedStack - The new fixed stack of Group elements.
     * @param {Array<Group>} newSlidingStack - The new sliding stack of Group elements.
     * @param {number} [slidingStart] - The starting position for sliding elements, optional.
     */
    drawStacks(newFixedStack, newSlidingStack, slidingStart) {
        const concat = (arrays) => [].concat(...arrays);

        for (let group of concat([this.oldFixedStack, this.oldSlidingStack])) {
            if (!newFixedStack.includes(group) && !newSlidingStack.includes(group)) {
                let editor = group.children[0];
                let row = editor.children[0];
                if (row.classList.contains('sticky-header')) {
                    row.style.top = '';
                    row.style.width = '';
                    row.classList.remove('sticky-header');
                    editor.removeChild(editor.children[1]);
                    if (this.DEBUG) console.log(`"${group.name.text}" removeAt(1)`);
                    row.classList.remove('sliding');
                }
            }
        }

        for (let group of concat([newFixedStack, newSlidingStack])) {
            let editor = group.children[0];
            let row = editor.children[0];
            if (!row.classList.contains('sticky-header')) {
                if (this.DEBUG) console.log(`"${group.name.text}" insertBefore(document.createElement(div), ...)`);
                let spacer = document.createElement('div');
                spacer.style.height = `${row.clientHeight}px`;
                row.style.width = `${row.clientWidth}px`;
                row.classList.add('sticky-header');
                editor.insertBefore(spacer, editor.children[1]);
            }
        }

        let offset = this.TOP - 1;
        for (let model of newFixedStack) {
            let editor = model.children[0];
            let row = editor.children[0];
            row.style.top = `${offset}px`;
            offset += row.clientHeight - 1;
            row.classList.toggle('sliding', false);
        }

        for (let model of newSlidingStack) {
            let editor = model.children[0];
            let row = editor.children[0];
            slidingStart -= row.clientHeight;
            row.style.top = `${slidingStart}px`;
            row.classList.toggle('sliding', true);
        }

        this.oldFixedStack = newFixedStack;
        this.oldSlidingStack = newSlidingStack;
    }
    /**
     * Finds the closest Group element in the ancestor hierarchy of a given element.
     * @param {HTMLElement} element - The starting element to search from.
     * @returns {Group|null} The found Group element, or null if none is found.
     */
    findGroup(element) {
        while (element) {
            if (element instanceof Group) { // Group is a class extending HTMLElement
                return element;
            }
            element = element.parentElement;
        }
        return null;
    }
    /**
     * Creates a stack of Group elements starting from a specified Group and moving up its parent chain.
     * @param {Group} group - The starting Group element.
     * @returns {Array<Group>} An array representing the stack of Group elements.
     */
    makeGroupStack(group) {
        let newStack = [];
        if (group !== null) {
            newStack.unshift(group); // unshift is used to insert at the beginning of the array
            while (group !== null && !(group instanceof Job)) { // Checking if it's not an instance of Job
                group = group.parentElement;
                if (group instanceof Group) {
                    newStack.unshift(group);
                }
            }
        }
        return newStack;
    }
}